### ISSUE TITLE
Need Help With the Arch Port

### ISSUE THREAD

**Issue by gardotd426** [regolith-linux/regolith-desktop#522]
So, hopefully you remember a while back when I started the project to get Regolith ported over to Arch. 

Well, it was working great, and I was gonna post it live to the AUR after getting enough people to test it (and I was pretty much ready), and all the sudden it broke. Well, I figured something happened with the move to GNOME 3.38. So, I looked, saw the new packages on regolith for 20.10, and decided to move over to those as the base. 

Well, that didn't fix it, but then I discovered something weird. So, the problem right now is that i3 loads, all the non-GNOME-control-center (and non logging-out) keybindings from the config work, the panel loads, etc., but there is no wallpaper, and gnome-control-center segfaults when you try to open it, and you can't log out with the keybinding. Well, if I open a terminal and type `gnome-flashback -r` (-r for replace), instantly the wallpaper pops up, the keybindings for the control center start working, everything works exactly as intended. 

Well, this confused me, so I started looking in /usr/bin/regolith-session and /usr/bin/regolith-session-init. Nothing I could find there that I can think of that causes it. 

But I also couldn't even find where /usr/bin/regolith-session-init gets called.... /usr/bin/regolith-session doesn't execute it, /usr/share/xsessions/regolith.desktop doesn't either. So how does /usr/bin/regolith-session-init get called upon? Because I also found out that running regolith-session-init from a terminal **after** (and only after) launching regolith from the display manager, it also has the same effect as `gnome-flashback` -r. 

Any help you could give me would be hugely appreciated. If you might know why the wallpaper and GNOME stuff isn't working, or if you could tell me a little bit about the init of the DE so I could try and just work it out myself, or whatever you're able to give. Thanks a lot. 

Oh and I'm happy to provide any info you need, I just don't know of any logs that would help or anything like that.

**Comment by kgilmer** [comment]
Hi @gardotd426, good to hear from you :smile: 

Yes, the relationship between the regolith Xsession and gnome-flashback is confusing.  In fact I didn't fully get it until a few versions after I released Regolith.  The connection is implicit, and spans a few various XDG file types.  The best way to understand how this works is to take a look at this project: https://github.com/deuill/i3-gnome-flashback

In addition to anything you may find helpful in the AUR, have a look at the link to the GNOME wiki.  This is the best documentation I have found that explains how gnome-flashback gets loaded from an Xsession file.  You may also find it helpful to look at the Metacity Xsession file that comes with gnome-flashback, as that also relies on gnome-session to invoke gnome-flashback from a different window manager.

Hope that helps and LMK if there are other questions I can answer!

**Comment by gardotd426** [comment]
Basically anything you can tell me about why this specific issue might be
happening (like the no wallpaper and gnome control center/logout not
working)

**Comment by gardotd426** [comment]
Also where regolith-session-init comes into play, because there's nothing I
see that calls it.

**Comment by kgilmer** [comment]
> Basically anything you can tell me about why this specific issue might be happening (like the no wallpaper and gnome control center/logout not working)

Like you have found, this is because gnome-flashback is not being loaded during session init.  This is done by `gnome-session`, described in the link I sent earlier.

> Also where regolith-session-init comes into play, because there's nothing I see that calls it.

See [here](https://github.com/regolith-linux/regolith-gnome-flashback/blob/master/usr/share/applications/regolith.desktop#L5).  This is also being loaded by `gnome-session`.

**Comment by gardotd426** [comment]
See that's what I'm confused about. I'll look into it more, but shouldn't
that be in /usr/share/xsessions? Anyway thanks for the help, I'll read all
the documentation and see what I can find out.

**Comment by kgilmer** [comment]
This is my understanding about what's going on:
1. OS boots, display manager (GDM or lightdm) loads
2. User logs in, specifying Regolith Xsession
3. Regolith xsession calls /usr/bin/regolith-session
4. regolith-session loads gnome-session
5. Via the params passed in (`exec gnome-session --builtin --session=regolith ...`) gnome-session calls regolith-session-init via the regolith desktop file (lined above)
6. regolith-session-init loads Xresources and starts i3

**Comment by gardotd426** [comment]
Okay that makes way more sense. It's so weird then that it's not working
over here. I'll see what else I can find, thanks so much.

**Comment by kgilmer** [comment]
Happy to help.  When troubleshooting situations like this I find it helpful to add lines in the shells scripts that append the current time to some file in /tmp.  Then I can see the order of execution and verifying that each script is running.  Something like `echo "$(date +%s) some script did something..." >> /tmp/debug.txt`

**Comment by gardotd426** [comment]
So funnily enough, I think I've figured it out. 

See, obviously Arch doesn't use dpkg, so in /usr/bin/regolith-session, these lines: 

```
# gnome-session broke compatability between 19.04 (3.32.0) and 19.10 (3.34.1) versions of Ubuntu (gnome-session-bin).
GNOME_SESSION_BIN_VERSION=$(dpkg -s gnome-session-bin | grep '^Version:' | cut -d' ' -f 2 | cut -d'-' -f 1)
COMPACT_VERSION="${GNOME_SESSION_BIN_VERSION//.}"

if [ $COMPACT_VERSION -gt 3320 ]; then
    exec gnome-session --builtin --session=regolith --disable-acceleration-check "$@"
else
    exec gnome-session --session=regolith --disable-acceleration-check "$@"
fi
```

had to be patched. Originally, back in Regolith 1.4, when it was called /usr/bin/i3-gnome-flashback, I was patching it to look like this:

```
    exec gnome-session --session=regolith --disable-acceleration-check "$@"
```
Because this is for Arch which would never have that old of a GNOME version. And it worked. Only now, it doesn't, and your comment here:

>Via the params passed in (exec gnome-session --builtin --session=regolith ...) gnome-session calls regolith-session-init via the regolith desktop file (lined above)

Gave me an idea. So I instead used the line with --builtin, and now it works. So `--builtin` is currently required (I don't know if there was a change to gnome-session over here or what to require `--builtin`, but now it's working). 

The only remaining issue is I've had reports of people's KB media keys not working when they say they do work in Regolith proper, so I have to look into that, but gnome-control-center, the wallpaper setting, logging out, all that stuff works as intended now. Thanks for your help, I'm getting close to being able to put it on the AUR.

**Comment by gardotd426** [comment]
Everything's working great, I've opened it up to mass testing by posting about it on r/archlinux on Reddit, and through the extra testing, I've found that regolith-look gives an error if gnome-terminal isn't installed (because it's trying to set a profile, etc.). What would be the best way in your opinion to check if gnome-terminal is installed and then not give the error if it's not present?

I was just thinking of something like:

```
if [ -n $(pacman -Qq | grep gnome-terminal) ]; then
UPDATE_TERM_FLAG=$(xrescat gnome.terminal.update true)                                                                  
  if [[ "$UPDATE_TERM_FLAG" == "true" && -f "/usr/bin/create-regolith-term-profile" ]] ; then                                 
  /usr/bin/create-regolith-term-profile
  fi
fi 
```
where regolith-look has:

```
 UPDATE_TERM_FLAG=$(xrescat gnome.terminal.update true)                                                                  
if [[ "$UPDATE_TERM_FLAG" == "true" && -f "/usr/bin/create-regolith-term-profile" ]] ; then
/usr/bin/create-regolith-term-profile
fi
```
So basically, that command only gets run if gnome-terminal is installed `pacman -Qq` lists packages installed on the system                                                                            

I'm not the greatest at if/thens (or bash scripting in general) yet, but I just figured this was the easiest way to do it without losing any functionality and without messing too much with the code. What do you think?

**Comment by gardotd426** [comment]
Or could I just do:

```
UPDATE_TERM_FLAG=$(xrescat gnome.terminal.update true)                                                                  
if [[ "$UPDATE_TERM_FLAG" == "true" && -f "/usr/bin/create-regolith-term-profile" && -n $(pacman -Qq | grep gnome-terminal) ]] ; then
/usr/bin/create-regolith-term-profile
fi
```
Or

```
UPDATE_TERM_FLAG=$(xrescat gnome.terminal.update true)                                                                  
if [[ "$UPDATE_TERM_FLAG" == "true" && -f "/usr/bin/create-regolith-term-profile" && -f /usr/bin/gnome-terminal ]] ; then
/usr/bin/create-regolith-term-profile
fi
```

**Comment by kgilmer** [comment]
Can you provide the exact error that you're seeing?

Ideally dependency on gnome-terminal would be expressed in the package dependency metadata.  There is no explicit dependency from `regolith-styles` (which ships `regolith-look`) to `gnome-terminal`.  (There is such a dependency from `regolith-i3-gaps-config` though)  So I can be certain the best way of fixing the issue but yeah it seems that `regolith-look` should not call the terminal update script if `gnome-terminal` is not resolved.  However, I'd like to do this without and package manager calls if possible, to make future porting easier.

**Comment by gardotd426** [comment]
Yeah sure, it's this error from `create-regolith-term-profile`, which you have set to be ran if those two conditions from /usr/bin/regolith-look are met. 

I went ahead and changed it to this:
```
if [[ "$UPDATE_TERM_FLAG" == "true" && -f "/usr/bin/create-regolith-term-profile" && -f "/usr/bin/gnome-terminal" ]] ; then
/usr/bin/create-regolith-term-profile
```
So it only runs if /usr/bin/gnome-terminal is present. Which is also without package managers. And, doing that did in fact fix the user's issue and the error no longer shows up. 

So yeah, basically, regolith-look is trying to run create-regolith-term-profile, but if gnome-terminal isn't installed, then create-regolith-term-profile will give that error, I just added a third condition that requires /usr/bin/gnome-terminal to *also* be there before create-regolith-term-profile is ran. It works over here. 

I've also found an error in your net-traffic i3xrocks bash script, I don't know if it only causes an error on Arch or what, but you're calling on the INT_TYPE to handle button events before actually *setting* INT_TYPE, so it won't work. All I did was literally just move the setting of INT_TYPE to before the button event handling, and it works perfectly now. 

See, you had this:

```
# Handle button event
if [ "x${BUTTON}" == "x1" ] ; then
  if [ "$INT_TYPE" == "W" ]; then
    ACTION=$(xrescat i3xrocks.action.net-traffic.left "/usr/bin/gnome-control-center --class=floating_window wifi")
    /usr/bin/i3-msg -q exec "$ACTION"
  else
    ACTION=$(xrescat i3xrocks.action.net-traffic.right "/usr/bin/gnome-control-center --class=floating_window network")
    /usr/bin/i3-msg -q exec "$ACTION"
  fi
fi
```

But the `INT_TYPE` isn't set yet. It's not set until later, here:

```
# determine the net interface name
IF=$(ip route get 1.1.1.1 | grep -Po '(?<=dev\s)\w+' | cut -f1 -d ' ')
  IF="${BLOCK_INSTANCE:-${IF}}"

    [ "$IF" = "" ] && exit

# set net icon based on wire or wireless connection
case "$IF" in
  e*)
    NIC=${label_icon:-$(xrescat i3xrocks.label.wired )}
    INT_TYPE="C"
    ;;
  w*)
    NIC=${label_icon:-$(xrescat i3xrocks.label.wifi )}
    INT_TYPE="W"
    ;;
esac
```

Well, I looked at your documentation on debugging i3xrocks, and watching /tmp/i3xrocks.err while I clicked told me that `INT_TYPE` was undefined, so I literally just switched the order, and now button events are handled properly.

**Comment by kgilmer** [comment]
Thanks for the details regarding the problems you've found.  I've updated `regolith-look` in a way similar to yours (I test for the existence of `gnome-terminal` on the path, rather than checking for the file in `/usr/bin` to cover more possible user environments.  Regarding the net_traffic bug, I believe this whole script has been rewritten since 1.5 has been released.  At least I do not see the issue in the current version in git.

You're very welcome to create PRs yourself for issues that you find @gardotd426 .  If you'd like me to mail you some Regolith stickers, email me a postal address and I'll send some to you.  International is OK.

**Comment by gardotd426** [comment]
That sounds great. I went ahead and changed my regolith-look patch to match up with your change, as I actually originally planned to do something like that but wasn't quite sure how best to do it, and so I checked Ubuntu-based distros at least to make sure gnome-terminal would also be in /usr/bin, and it was, so I figured that was good enough lol. But yeah this is much better.

>Regarding the net_traffic bug, I believe this whole script has been rewritten since 1.5 has been released. At least I do not see the issue in the current version in git.

Yeah, I just now looked at your -git version of net-traffic and you indeed have it the same way I have it now (with INT_TYPE being set before the button event handling), so you inadvertently fixed that as well. I might move over to the -git i3xrocks scripts instead of using the release .deb packages since it's just little script files and would be pretty straightforward. These look like the best available versions of these scripts. 

And yeah I'll be happy to file any PRs if I find any more issues, I just didn't wanna blow up your spot or anything, especially since I'm trying to do something some people might consider "unsupported." But I'm super pumped to have your support, it's really cool.
